doc/api/fs.md

```diff
@@ -1308,6 +1308,38 @@ characters directly to the `prefix` string. For instance, given a directory
 `prefix` must end with a trailing platform-specific path separator
 (`require('node:path').sep`).
 
+### `fsPromises.mkdtempDisposable(prefix[, options])`
+
+<!-- YAML
+added: TODO
+changes:
+  - version:
+      - TODO
+    pr-url: TODO
+    description: Initial version.  <-- this doesn't seem to be common practice but the result is that there is no pr-url for the initial feature
+-->
+
+* `prefix` {string|Buffer|URL}
+* `options` {string|Object}
+  * `encoding` {string} **Default:** `'utf8'`
+* Returns: {Promise} Fulfills with a Promise for an async-disposable object with a "path" property.
+
+The resulting Promise holds an async-disposable object whose `path` property
+holds the created directory path. When the object is disposed, the directory
+and its contents will be removed asynchronously. The object has an async
+`remove()` method which will perform the same task.
+
+Both this function and the disposal function on the resulting object are
+async, so it should be used with `await` + `await using` as in
+`await using dir = await fsPromises.mkdtempDisposable('prefix')`.
```

Comment:
The double await here just *feels* awkward. It occurs to me that instead of having this method return a promise we could have it return a custom thenable that can act as both the promise and the disposer. I'm not sure if that's a good idea or an absolutely terrible one.

```
await using dir = fsPromises.mktempDisposable('prefix');
console.log(await dir);  // works just like if it were a promise...
```

Reply:
The 2x2 matrix of sync/async initialisation and sync/async disposal is just an intrinsic part of the language feature (and then there's [this](https://github.com/tc39/proposal-explicit-resource-management?tab=readme-ov-file#await-using-in-for-of-and-for-await-of-statements), if you have an aspirin handy...) It feels weird, granted, but I imagine it'll bed in over time.

This suggested paradigm seems like it could really run into trouble – not only is it difficult to tell at a glance how to access the result, but there is also the possibility of a disposer existing, and being invoked, before its corresponding resource has been initialised.

Reply:
Yeah I don't like the syntax here either, but it is what we (TC39) ended up going with, with the expectation that it would look like this. I don't think it's worth trying to be clever to get around the awkwardness which is built into the language now.